### DISCUSSION TITLE
Unknown file extension \".ts\"

### DISCUSSION THREAD

**Discussion by imcm7** [mikro-orm/mikro-orm#4727] [General]
Unknown file extension \".ts\" for /home/nuxt/server/entities/user.entity.ts"

```
const options = defineConfig({
  entities: ['./dist/entities/**/*.js'],
  entitiesTs: ['./server/entities/*.ts'], // path to our TS entities (source), relative to `baseDir`
});
```

Why it not auto import it and I always need to enumerate it or to use 
```
import * as entities from '~/server/entities/index';
const options = defineConfig({
  entities: Object.values(entities),
  entitiesTs: Object.values(entities), // path to our TS entities (source), relative to `baseDir`
});

**Comment by B4nan** [comment]
Sounds like the same error that is mentioned in [the guide](https://github.com/mikro-orm/guide), if so, try adding this to your ORM config:

```ts
dynamicImportProvider: id => import(id),
```

Next time I would greatly appreciate it if you start providing more details right ahead, it's really hard to help when I have no idea on your project setup. The very basics you should always provide: full error message with stack trace, whether your project is ESM/CJS, what your ORM config is and what version are you on.

>  entities: ['./dist/entities/**/*.js'],
>  entitiesTs: ['./server/entities/*.ts'], // path to our TS entities (source), relative to `baseDir`

FYI this seems misconfigured, you should always have the both paths matching, and those are not - `./dist/entities/**/*.js` will resolve to any JS file in any folder under the `dist/entities`, while `./server/entities/*.ts` only matches TS files directly in the `server/entities` (also, do you really compile from `server` directory? no `src` folder?)

**Reply by imcm7** [reply]
That `./dist/entities/**/*.js` will be used in prod and now is used `entitiesTs`? 
Now I don't know what path I will put in `entities`.

**Reply by imcm7** [reply]
No src dir in nuxt https://github.com/imcm7/nuxt-db

**Reply by imcm7** [reply]
Not help `dynamicImportProvider: id => import(id),` and 

```
"ts-node": {
    "esm": true,
    "transpileOnly": true
  }
  ```

**Reply by imcm7** [reply]
If I run `mikro-orm-esm debug` 

```
 could use `entities` array (contains 0 references and 1 paths)
   - /home/nuxt/entities/*.js (not found)
 - will use `entitiesTs` array (contains 0 references and 1 paths)
   - /home/nuxt/server/entities/*.ts (found)

```

**Reply by imcm7** [reply]
> FYI this seems misconfigured, you should always have the both paths matching, and those are not

Prior to ask you I try different paths, and the same paths except extension too.
But I think the paths will differ because nux compile for prod server to `.nuxt` dir.

**Reply by imcm7** [reply]
May be is need to add `metadataProvider: TsMorphMetadataProvider,`

**Reply by imcm7** [reply]
```
const options = defineConfig({
  // optionally you can override the base directory (defaults to `process.cwd()`)
  allowGlobalContext: false,
  baseDir: process.cwd(),
  dbName: 'nuxt',
  debug: true, // ['discovery', 'info', 'query', 'query-params'],
  dynamicImportProvider: id => import(id),
  discovery: {
    // disableDynamicFileAccess: true,
    warnWhenNoEntities: true
  },
  driver: MariaDbDriver,
  // entities: Object.values(entities),
  // entitiesTs: Object.values(entities),
  // entities: ['./server/entities'],
  // entitiesTs: ['./server/entities/*'],
  entities: ['./server/entities/*.js'], // path to our JS entities (dist), relative to `baseDir`
  entitiesTs: ['./server/entities/*.ts'], // path to our TS entities (source), relative to `baseDir`
  extensions: [Migrator],
  loadStrategy: LoadStrategy.JOINED,
  metadataProvider: TsMorphMetadataProvider,
  migrations: {
    tableName: 'mikro_orm_migrations', // migrations table name
    path: process.cwd() + '/migrations', // path to folder with migration files
    glob: '!(*.d).{js,ts}', // how to match migration files (all .js and .ts files, but not .d.ts)
    transactional: true, // run each migration inside transaction
    disableForeignKeys: true, // try to disable foreign_key_checks (or equivalent)
    allOrNothing: true, // run all migrations in current batch in master transaction
    emit: 'ts', // migration generation mode
  },
  password: 'changeme',
  populateAfterFlush: true,
  port: 3306,
  // registerRequestContext: false,
  // seeder: {
  //   path: './seeders',
  //   defaultSeeder: 'DatabaseSeeder',
  // },
  tsNode: true,
  user: 'root'
});
```

**Reply by imcm7** [reply]
```
[discovery] ORM entity discovery started, using TsMorphMetadataProvider
[nuxt] [request error] [unhandled] [500] Unknown file extension ".ts" for /home/nuxt/server/entities/user.entity.ts
  at __node_internal_captureLargerStackTrace (node:internal/errors:496:5)  
  at new NodeError (node:internal/errors:405:5)  
  at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:99:9)  
  at defaultGetFormat (node:internal/modules/esm/get_format:142:36)  
  at defaultLoad (node:internal/modules/esm/load:115:20)  
  at ModuleLoader.load (node:internal/modules/esm/loader:388:13)  
  at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:270:56)  
  at new ModuleJob (node:internal/modules/esm/module_job:65:26)  
  at #createModuleJob (node:internal/modules/esm/loader:282:17)  
  at ModuleLoader.getJobFromResolveResult (node:internal/modules/esm/loader:240:34)  
  at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:221:17)  
  at async ModuleLoader.import (node:internal/modules/esm/loader:307:23)
[discovery] - processing 6 files

```

**Reply by B4nan** [reply]
Looked at your repro, and the paths are definitely misconfigured, nuxt is bundling everything to a single file and (as expected) it does not generate the `.d.ts` files - so ts-morph won't work for production builds. I can also see that nuxt is apparently not bringing any kind of TS support, instead, it always compiles the whole thing and runs the compiled JS (again, to the `.nuxt` folder, so even the source paths were wrong, not just the dist ones).

Using the entity class references (so `entities: [User, ...]`) is the best thing you can do in such constraints. I will think about this more when I have a moment, it would be great to come up with more vite native support, I know this can be confusing.
